File: app/src/main/java/net/mbonnin/arcanetracker/adapter/DetailsView.java
package net.mbonnin.arcanetracker.adapter;

import android.content.Context;
import android.graphics.Bitmap;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;

import net.mbonnin.arcanetracker.CardRenderer;
import net.mbonnin.arcanetracker.CardUtil;
import net.mbonnin.arcanetracker.R;
import net.mbonnin.arcanetracker.Typefaces;
import net.mbonnin.arcanetracker.Utils;
import net.mbonnin.arcanetracker.databinding.DetailsViewBinding;
import net.mbonnin.arcanetracker.parser.Entity;
import net.mbonnin.arcanetracker.parser.GameLogic;
import net.mbonnin.hsmodel.CardId;
import net.mbonnin.hsmodel.PlayerClass;
import net.mbonnin.hsmodel.Type;

import java.util.ArrayList;
import java.util.List;

public class DetailsView extends LinearLayout {
    private int mTopMargin;
    private int mCardWidth;

    public DetailsView(Context context) {
        super(context);
        setOrientation(HORIZONTAL);

    }

    public void configure(Bitmap bitmap, DeckEntryItem deckEntryItem, int height) {

        int w = (height * CardRenderer.TOTAL_WIDTH) / CardRenderer.TOTAL_HEIGHT;
        if (bitmap != null) {
            ImageView imageView = new ImageView(getContext());
            imageView.setImageBitmap(bitmap);
            LayoutParams layoutParams = new LayoutParams(w, height);
            addView(imageView, layoutParams);
        }

        mCardWidth = w;
        mTopMargin = 30;

        for (Entity entity : deckEntryItem.entityList) {
            DetailsViewBinding b = DetailsViewBinding.inflate(LayoutInflater.from(getContext()));

            StringBuilder builder = new StringBuilder();

            if (Utils.INSTANCE.isAppDebuggable()) {
                builder.append(getContext().getString(R.string.card, entity.EntityID));
                builder.append("\n");
            }


            String cardType = entity.tags.get(Entity.KEY_CARDTYPE);
            if (entity.extra.drawTurn != -1) {
                builder.append(getContext().getString(R.string.drawnTurn, GameLogic.gameTurnToHumanTurn(entity.extra.drawTurn)));
                if (entity.extra.mulliganed) {
                    builder.append(" (");
                    builder.append(getContext().getString(R.string.mulliganed));
                    builder.append(")");
                }
                builder.append("\n");
            }

            if (entity.extra.playTurn != -1) {
                builder.append(getContext().getString(R.string.playedTurn, GameLogic.gameTurnToHumanTurn(entity.extra.playTurn)));
                builder.append("\n");
            }
            if (entity.extra.diedTurn != -1 && (Type.MINION.equals(cardType) || Type.WEAPON.equals(cardType))) {
                builder.append(getContext().getString(R.string.diedTurn, GameLogic.gameTurnToHumanTurn(entity.extra.diedTurn)));
                builder.append("\n");
            }
            if (!TextUtils.isEmpty(entity.extra.createdBy)) {
                builder.append(getContext().getString(R.string.createdBy, CardUtil.INSTANCE.getCard(entity.extra.createdBy).name));
            }

            if (Entity.ZONE_SECRET.equals(entity.tags.get(Entity.KEY_ZONE))
                    && TextUtils.isEmpty(entity.CardID)) {
                builder.append(Utils.INSTANCE.getString(R.string.possibleSecrets));
                appendPossibleSecrets((LinearLayout) b.getRoot(), entity);
            }

            String s = builder.toString();

            if (Utils.INSTANCE.isEmpty(s)) {
                builder.append(Utils.INSTANCE.getString(R.string.inDeck));
                builder.append("\n");
                s = builder.toString();
            }

            b.textView.setText(s);
            b.textView.setTypeface(Typefaces.franklin());

            addView(b.getRoot());
        }

        applyMargins();
    }

    void applyMargins() {
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            if (child instanceof ImageView) {
                continue;
            }
            LayoutParams layoutParams = new LayoutParams(mCardWidth, ViewGroup.LayoutParams.WRAP_CONTENT);
            int p = Utils.INSTANCE.dpToPx(5);
            layoutParams.leftMargin = layoutParams.rightMargin = p;
            layoutParams.topMargin = Utils.INSTANCE.dpToPx(mTopMargin);
            child.setLayoutParams(layoutParams);
        }
        requestLayout();
    }

    public void setTopMargin(int topMargin) {
        mTopMargin = topMargin;
        applyMargins();
    }

    private void appendPossibleSecrets(LinearLayout verticalLayout, Entity entity) {
        String playerClass = entity.tags.get(Entity.KEY_CLASS);

        if (playerClass == null) {
            return;
        }

        List<DeckEntryItem> list = new ArrayList<>();

        switch (playerClass) {
            case PlayerClass.HUNTER:
                addSecret(list, CardId.BEAR_TRAP, entity.extra.selfHeroAttacked);
                addSecret(list, CardId.CAT_TRICK, entity.extra.otherPlayerCastSpell);
                addSecret(list, CardId.EXPLOSIVE_TRAP, entity.extra.selfHeroAttacked);
                addSecret(list, CardId.FREEZING_TRAP, entity.extra.selfHeroAttacked || entity.extra.selfMinionWasAttacked);
                addSecret(list, CardId.SNAKE_TRAP, entity.extra.selfMinionWasAttacked);
                addSecret(list, CardId.SNIPE, entity.extra.otherPlayerPlayedMinion);
                addSecret(list, CardId.DART_TRAP, entity.extra.otherPlayerHeroPowered);
                addSecret(list, CardId.HIDDEN_CACHE, entity.extra.otherPlayerPlayedMinion);
                addSecret(list, CardId.MISDIRECTION, entity.extra.selfHeroAttacked);
                addSecret(list, CardId.VENOMSTRIKE_TRAP, entity.extra.selfMinionWasAttacked);
                break;
            case PlayerClass.MAGE:
                addSecret(list, CardId.MIRROR_ENTITY, entity.extra.otherPlayerPlayedMinion);
                addSecret(list, CardId.MANA_BIND, entity.extra.otherPlayerCastSpell);
                addSecret(list, CardId.COUNTERSPELL, entity.extra.otherPlayerCastSpell);
                addSecret(list, CardId.EFFIGY, entity.extra.selfPlayerMinionDied);
                addSecret(list, CardId.ICE_BARRIER, entity.extra.selfHeroAttacked);
                addSecret(list, CardId.POTION_OF_POLYMORPH, entity.extra.otherPlayerPlayedMinion);
                addSecret(list, CardId.DUPLICATE, entity.extra.selfPlayerMinionDied);
                addSecret(list, CardId.VAPORIZE, entity.extra.selfHeroAttackedByMinion);
                addSecret(list, CardId.ICE_BLOCK, false);
                addSecret(list, CardId.SPELLBENDER, entity.extra.selfMinionTargetedBySpell);
                addSecret(list, CardId.FROZEN_CLONE, entity.extra.otherPlayerPlayedMinion);
                break;
            case PlayerClass.PALADIN:
                addSecret(list, CardId.COMPETITIVE_SPIRIT, entity.extra.competitiveSpiritTriggerConditionHappened);
                addSecret(list, CardId.AVENGE, entity.extra.selfPlayerMinionDied);
                addSecret(list, CardId.REDEMPTION, entity.extra.selfPlayerMinionDied);
                addSecret(list, CardId.REPENTANCE, entity.extra.otherPlayerPlayedMinion);
                addSecret(list, CardId.SACRED_TRIAL, entity.extra.otherPlayerPlayedMinionWithThreeOnBoardAlready);
                addSecret(list, CardId.NOBLE_SACRIFICE, entity.extra.selfHeroAttacked || entity.extra.selfMinionWasAttacked);
                addSecret(list, CardId.GETAWAY_KODO, entity.extra.selfPlayerMinionDied);
                addSecret(list, CardId.EYE_FOR_AN_EYE, entity.extra.selfHeroAttacked);
                break;
        }

        int i = 0;
        LinearLayout horizontalLayout = null;
        for (DeckEntryItem deckEntryItem: list) {
            if (i%2 == 0) {
                horizontalLayout = new LinearLayout(getContext());
                horizontalLayout.setOrientation(HORIZONTAL);
                //verticalLayout.addView(horizontalLayout, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, Utils.dpToPx(30)));
            }

            View view = LayoutInflater.from(getContext()).inflate(R.layout.bar_card, null);
            ViewGroup barTemplate = (ViewGroup) LayoutInflater.from(getContext()).inflate(R.layout.bar_template, null);
            ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
            barTemplate.addView(view, 0, params);

            DeckEntryHolder holder = new DeckEntryHolder(barTemplate);
            holder.bind(deckEntryItem);

            verticalLayout.addView(barTemplate, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        }
    }

    private void addSecret(List<DeckEntryItem> list, String cardId, boolean condition) {
        DeckEntryItem deckEntryItem = new DeckEntryItem();
        deckEntryItem.card = CardUtil.INSTANCE.getCard(cardId);
        deckEntryItem.count = condition ? 0 : 1;
        list.add(deckEntryItem);
    }
}

File: app/src/main/java/net/mbonnin/arcanetracker/adapter/DetailsView.kt
package net.mbonnin.arcanetracker.adapter

import android.content.Context
import android.graphics.Bitmap
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.core.view.size
import net.mbonnin.arcanetracker.*
import net.mbonnin.arcanetracker.databinding.DetailsViewBinding
import net.mbonnin.arcanetracker.parser.Entity
import net.mbonnin.arcanetracker.parser.GameLogic
import net.mbonnin.hsmodel.enum.Type

class DetailsView(context: Context) : LinearLayout(context) {
    private var mTopMargin: Int = 0
    private var mCardWidth: Int = 0

    init {
        orientation = LinearLayout.HORIZONTAL

    }

    fun configure(bitmap: Bitmap?, entityList: List<Entity>, height: Int) {

        val w = height * CardRenderer.TOTAL_WIDTH / CardRenderer.TOTAL_HEIGHT
        if (bitmap != null) {
            val imageView = ImageView(context)
            imageView.setImageBitmap(bitmap)
            val layoutParams = LinearLayout.LayoutParams(w, height)
            addView(imageView, layoutParams)
        }

        mCardWidth = w
        mTopMargin = 30

        for (entity in entityList) {
            val b = DetailsViewBinding.inflate(LayoutInflater.from(context))
            b.root.minimumWidth = mCardWidth

            val builder = StringBuilder()

            if (Utils.isAppDebuggable) {
                builder.append(context.getString(R.string.card, entity.EntityID))
                builder.append("\n")
            }


            val cardType = entity.tags[Entity.KEY_CARDTYPE]
            if (entity.extra.drawTurn != -1) {
                builder.append(context.getString(R.string.drawnTurn, GameLogic.gameTurnToHumanTurn(entity.extra.drawTurn)))
                if (entity.extra.mulliganed) {
                    builder.append(" (")
                    builder.append(context.getString(R.string.mulliganed))
                    builder.append(")")
                }
                builder.append("\n")
            }

            if (entity.extra.playTurn != -1) {
                builder.append(context.getString(R.string.playedTurn, GameLogic.gameTurnToHumanTurn(entity.extra.playTurn)))
                builder.append("\n")
            }
            if (entity.extra.diedTurn != -1 && (Type.MINION == cardType || Type.WEAPON == cardType)) {
                builder.append(context.getString(R.string.diedTurn, GameLogic.gameTurnToHumanTurn(entity.extra.diedTurn)))
                builder.append("\n")
            }
            if (!TextUtils.isEmpty(entity.extra.createdBy)) {
                builder.append(context.getString(R.string.createdBy, CardUtil.getCard(entity.extra.createdBy!!).name))
            }

            if (Entity.ZONE_SECRET == entity.tags[Entity.KEY_ZONE] && entity.CardID.isNullOrEmpty()) {
                builder.append(Utils.getString(R.string.possibleSecrets))
                appendPossibleSecrets(b.secrets, entity)
            }

            var s = builder.toString()

            if (Utils.isEmpty(s)) {
                builder.append(Utils.getString(R.string.inDeck))
                builder.append("\n")
                s = builder.toString()
            }

            b.textView.text = s
            b.textView.typeface = Typefaces.franklin()

            addView(b.root)
        }

        applyMargins()
    }

    internal fun applyMargins() {
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child is ImageView) {
                continue
            }
            val layoutParams = LinearLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT)
            val p = Utils.dpToPx(5)
            layoutParams.rightMargin = p
            layoutParams.leftMargin = layoutParams.rightMargin
            layoutParams.topMargin = Utils.dpToPx(mTopMargin)
            child.layoutParams = layoutParams
        }
        requestLayout()
    }

    fun setTopMargin(topMargin: Int) {
        mTopMargin = topMargin
        applyMargins()
    }

    private fun appendPossibleSecrets(horizontalLayout: LinearLayout, entity: Entity) {
        val game = GameLogicListener.get().currentGame

        val possibleSecrets: Collection<String>

        if (TestSwitch.SECRET_LAYOUT) {
            possibleSecrets = CardUtil.possibleSecretList(entity.tags[Entity.KEY_CLASS], GameType.GT_RANKED.name, FormatType.FT_WILD.name)
        } else {
            if (game == null) {
                return
            }
            possibleSecrets = CardUtil.possibleSecretList(entity.tags[Entity.KEY_CLASS], game.gameType, game.formatType)
        }

        val list = possibleSecrets.map {
            val deckEntryItem = DeckEntryItem(card = CardUtil.getCard(it))
            deckEntryItem.count = if (entity.extra.excludedSecretList.contains(it)) 0 else 1

            deckEntryItem
        }

        var verticalLayout: LinearLayout? = null
        for (deckEntryItem in list) {

            if (verticalLayout == null) {
                verticalLayout = LinearLayout(horizontalLayout.context)
                verticalLayout.orientation = VERTICAL
                horizontalLayout.addView(verticalLayout, mCardWidth - Utils.dpToPx(40), WRAP_CONTENT)
            }
            val view = LayoutInflater.from(context).inflate(R.layout.bar_card, null)
            val barTemplate = LayoutInflater.from(context).inflate(R.layout.bar_template, null) as ViewGroup
            val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            barTemplate.addView(view, 0, params)

            val holder = DeckEntryHolder(barTemplate)
            holder.bind(deckEntryItem)

            verticalLayout.addView(barTemplate, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, Utils.dpToPx(30)))

            if (verticalLayout.size >= 8) {
                // if we have a lot of secrets, they won't fit on a single screen
                verticalLayout = null
            }
        }
    }
}
